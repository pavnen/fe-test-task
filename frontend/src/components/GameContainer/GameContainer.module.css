.gameContainerHeader {
  font-size: 24px;
  margin-bottom: 20px;
}
