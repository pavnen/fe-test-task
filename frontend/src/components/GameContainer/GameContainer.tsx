import React, {useCallback, useState} from 'react'
import {Container, Toast} from 'react-bootstrap'
import {PageSwitcher} from '../PageSwitcher/PageSwitcher'
import {ScorePage} from '../ScorePage/ScorePage'
import {GamePage} from '../GamePage/GamePage'
import {api, GameResponse, ScoreResponse} from '../../api'

import styles from './GameContainer.module.css'

export enum PageToShow {
  Game = 'game',
  Score = 'score'
}

export const GameContainer: React.FC = () => {
  const [pageToShow, setPageToShow] = useState<PageToShow>(PageToShow.Game)
  const [game, setGame] = useState<GameResponse | null>(null)
  const [score, setScore] = useState<ScoreResponse | null>(null)
  const [alert, setAlert] = useState<string | undefined>()

  const closeAlert = useCallback(() => setAlert(undefined), [])

  const updateGame = (gameResponse: GameResponse | void) => {
    if (gameResponse) {
      setGame(gameResponse)
    }
  }

  const getGame = useCallback(() => api.game().then(updateGame), [])

  const reset = useCallback(() => api.reset().then(updateGame), [])

  const errorHandler = (err: Error) => setAlert(err.message || 'Server error...')

  const move = useCallback((cell: number) => api.move(cell).then(updateGame).catch(errorHandler), [])

  const getScore = useCallback(
    () =>
      api
        .score()
        .then(res => res && setScore(res))
        .catch(errorHandler),
    []
  )

  const pageDescriptions = [
    {title: 'game', callback: () => setPageToShow(PageToShow.Game)},
    {title: 'score', callback: () => setPageToShow(PageToShow.Score)}
  ]

  return (
    <Container>
      {alert && (
        <Toast autohide className={styles.toast} delay={3000} onClose={closeAlert}>
          <Toast.Header>Error!</Toast.Header>
          <Toast.Body>{alert}</Toast.Body>
        </Toast>
      )}
      <header className={styles.gameContainerHeader}>The tic-tac-toe game</header>
      <PageSwitcher currentPage={pageToShow} pages={pageDescriptions} />
      {pageToShow === PageToShow.Game ? (
        <GamePage game={game} getBoardCallback={getGame} moveCallback={move} resetCallback={reset} />
      ) : (
        <ScorePage score={score} getScore={getScore} />
      )}
    </Container>
  )
}
