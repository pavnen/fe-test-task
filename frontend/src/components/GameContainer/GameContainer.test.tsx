import React from 'react'
import {render} from '@testing-library/react'
import {GameContainer} from './GameContainer'

it('should match snapshot', () => {
  const {container} = render(<GameContainer />)
  expect(container).toMatchSnapshot()
})
