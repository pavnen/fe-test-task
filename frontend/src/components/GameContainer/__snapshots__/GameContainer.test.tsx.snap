// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
<div>
  <div
    class="container"
  >
    <header
      class="gameContainerHeader"
    >
      The tic-tac-toe game
    </header>
    <div
      class="pageSwitcherWrapper btn-group btn-group-lg"
      role="group"
    >
      <button
        class="pageSwitcherButton btn active btn-secondary"
        type="button"
      >
        game
      </button>
      <button
        class="pageSwitcherButton btn btn-secondary"
        type="button"
      >
        score
      </button>
    </div>
    <div>
      'Waiting for board...'
    </div>
  </div>
</div>
`;
